README.md
# ibmrcms




IBM RCMS & Retain System Terminal Tools for Macos

---------------
FUNCTION I. CR
---------------
1. Call Read
    Usage:  cr,<Call No.>
2. Call Read every 30 seconds waiting for parts PO
    Usage:  cr,<Call No.>,w

---------------
FUNCTION II. CS
---------------
1. Call Search for current login user
    Usage: cs
2. Call Search for user specified
    Usage: cs,<Notes ID>
3. Type of Call Search for user specified
    Usage: cs,<Notes ID>,CRR|INS|MES|PMA
4. IPS Call Search for user specified
    Usage: cs,SDLINC,<Notes ID>

----------------
FUNCTION III. CI
----------------
Call Inventory
    Usage: ci,<Cust No.>,<Machine SN>,<Machine Type>
    All 3 args or few of them.

---------------
FUNCTION IV. IN
---------------
Product Inventory
    Usage: in,<Machine Type>,<Machine SN>,<Cust No.>
    All 3 args or fre of them.

---------------
FUNCTION V. RED
---------------
1. Red Time Reporting List for Current User
    Usage: red
2. Summary for Current User of this Cutoff
    Usage: red,s
3. Backup Time Reporting Records to Excel File
    Usage: red,b
4. Red or Summary or Backup for Other Users
    Usage: red,AVR71D | red,AVR71D,s | red,AVR71d,b


----------------
FUNCTION VI. PMH
----------------
1. Read Open Problem for Current User
    Usage: z
2. Read Problem
    Usage: z,49123,000,672
3. Search Cust No.
    Usage: z,805543 | z,0805543 | z,00805543
4. Search Cust Name
    Usage: z,Gansu ICBC
5. Search Retain User
    Usage: z,Zhang Zhi Jie,open

    All These Args or Few of Them.
-------------------------------
1. Display User Profile
    Usage: dr,149715
    <Only Support Retain ID>

-------------------------------
1. Create Problem
    Usage: ce,<Machine Type>,<Machine SN>
2. Create Problem with RCMS Call No. Specified
    Usage: ce,call,P462YK9
    <Not supported yet.>

-----------------
NECESSARY PACKAGE
-----------------
1. python3 for MacOS or Windows
# Download from website and double click install

2. s3270 for MacOS/Linux or ws3270 for Windows
# Download from website, compile & install.
# Or install via homebrew

3. openpyxl for python3
# pip3 install openpyxl

Start Program:
# python3 rcmsMenu.py




CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [norvyn.com](https://norvyn.com).

## [Unreleased]
## [3.5] -2021-04-07
### Fixed
- Fix Cut-Off start from Sunday instead of Monday.

## [3.4] -2021-03-01
### Changed
- 'IN' command can read machine contract start/stop date

## [3.3] -2021-02-22
### Changed
- Change main menu style

### Add
- Add 'w' to find out SSR information,
Not supported yet.

## [3.2] - 2021-02-21
### Fixed
- Fix available hours count mistake between cut-off cycle

## [3.1] - 2021-02-21
### Added
- Port specified move to the configure file, so you can connect to the host via port forward.
- Add visible process when running 'red'.
- Add Changelog to record code release.
- Once PMH has been read, just type 'z' to repeat the action.

### Changed
- Move log files to dir 'log'
- Use 'ce' to create PMH only need machine SN

### Fixed
- Fix 'red' function does not fit for year 2021
- replace('2020', '20') --> replace('20', '', 1')
- Fix machine model not been read when use 'ce' to create a PMH 

### Removed
- Nothing removed.
<include config files>

[Unreleased]: https://norvyn.com
[3.3]: https://norvyn.com
[3.2]: https://norvyn.com
[3.1]: https://norvyn.com

functions/readConfig.py
from datetime import date
import os

conFilesDir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../config'))
# os.chdir(conFilesDir)
rcmsArgs = ['host1', 'host2', 'ssrSN', 'passwd', 'epsbSSRList']
rcmsConf = 'rcms.conf'
rcmsValues = {}

cutoffConf = 'cutoffDate.conf'
dateList = []
cutoffDate = []

teamSSRListConf = 'teamSSRList.conf'
teamSSRList = []

retainConf = 'retain.conf'
retainValues = {}
retainArgs = ['host1', 'host2', 'retainID', 'passwd', 'entryCmd', 'locQueue']


def getRCMSConf():
    if not os.path.isfile(conFilesDir + '/' + rcmsConf):
        raise FileNotFoundError('RCMS Config File Not Exist.')
    with open(conFilesDir + '/' + rcmsConf, 'r') as f:
        for dataLine in f.readlines():
            args = dataLine.split('=')[0].strip()
            if args in rcmsArgs:
                value = dataLine.split('=')[1].strip().strip("'").strip('"')
                rcmsValues[args] = value
        f.close()
    return rcmsValues


def getCutoffDate():
    if not os.path.isfile(conFilesDir + '/' + cutoffConf):
        raise FileNotFoundError('Cutoff Config File Not Exist.')
    with open(conFilesDir + '/' + cutoffConf, 'r') as f:
        for dataLine in f.readlines():
            if dataLine.startswith('20'):
                dateList.append(dataLine.replace('-', '').rstrip('\n'))
        currTimeMap = str(date.today()).replace('-', '')
        dateList.append(currTimeMap)
        dateList.sort()
        currDayIndex = dateList.index(currTimeMap)
        if int(dateList[currDayIndex]) - int(dateList[currDayIndex-1]) < 7:
            print('Less than 7 days since last cutoff.')
            print('Show the last cutoff cycle.')
            cutoffDate.append(dateList[currDayIndex-2])
            cutoffDate.append(dateList[currDayIndex-1])
        else:
            print('More than 7 days since last cutoff.')
            print('Show the current cutoff cycle.')
            cutoffDate.append(dateList[currDayIndex-1])
            cutoffDate.append(dateList[currDayIndex+1])
        f.close()
    return cutoffDate


def getTeamSSRList():
    if not os.path.isfile(conFilesDir + '/' + teamSSRListConf):
        raise FileNotFoundError('Team SSR List Not Exist.')
    with open(conFilesDir + '/' + teamSSRListConf, 'r') as f:
        for ssrList in f.readlines():
            if ssrList and not ssrList.startswith('#'):
                teamSSRList.append(ssrList.strip('\n').strip())
        f.close()
    return teamSSRList


def getRetainConf():
    if not os.path.isfile(conFilesDir + '/' + retainConf):
        raise FileNotFoundError('RETAIN Config File Not Exist.')
    with open(conFilesDir + '/' + retainConf, 'r') as f:
        for dataLine in f.readlines():
            args = dataLine.split('=')[0].strip()
            if args in retainArgs:
                value = dataLine.split('=')[1].strip().strip("'").strip('"')
                retainValues[args] = value
        f.close()
    return retainValues


if __name__ == '__main__':
    print(getRCMSConf())
    print(getRCMSConf()['host1'])
    print(getCutoffDate())
    print(getTeamSSRList())
    print(getRetainConf())



functions/readOTExcel_t.py
from functions import read_config
import os
from openpyxl import load_workbook
from datetime import date
import datetime
from functions import transform_rcms_date
import time

ot_files_dir = ''
ot_files = []
ot_items = []

try:
    ot_files_dir = read_config.ot_file_loc
    if ot_files_dir == '':
        raise RuntimeWarning('OT Files Not Defined!')
    for file in os.listdir(ot_files_dir):
        if file.endswith('xlsx') or file.endswith('xls'):
            if not file.startswith('~') and not file.startswith('.'):
                ot_files.append(file)
except RuntimeError:
    print('Read OT Request Form Location Failed!')
    
try:
    os.chdir(ot_files_dir)
except IsADirectoryError:
    print("'{0}' May not Exist.".format(ot_files_dir))

for file in ot_files:
    workbook = load_workbook(file, read_only=True, data_only=True)
    sheet = workbook.get_sheet_by_name('Sheet1')
    max_row = sheet.max_row
    row = 3
    while True:
        row += 1
        if row > max_row:
            break
        if not str(sheet.cell(row, 6).value).startswith('20'):
            continue
        ot_date = sheet.cell(row, 6).value
        ot_start_time = sheet.cell(row, 7).value
        ot_end_time = sheet.cell(row, 8).value
        ot_time = sheet.cell(row, 9).value
        ot_customer = sheet.cell(row, 4).value
        ot_content = sheet.cell(row, 5).value
        ot_items.append([ot_date, ot_start_time, ot_end_time, ot_time, ot_customer, ot_content])

# Change time format
# 18:00 to 180, 18:30 to 185
for line in ot_items:
    try:
        line[0] = line[0].strftime('%y%m%d')
        line[1] = line[1].strftime('%H') + str(int(line[2].strftime('%M')) // 6)
        line[2] = line[2].strftime('%H') + str(int(line[2].strftime('%M')) // 6)
        if line[2] == '000':
            line[2] = '240'
        if type(line[3]) == datetime.time:
            line[3] = '{:0>3d}'.format(int(line[3].strftime('%H')) * 10)
        else:
            raise ValueError('Wrong Format of Date in OT Excel!')
    except ValueError:
        print('Wrong Format of Date in OT Excel!')
        
ot_items.sort()


def get_ot_cycle(ot_month=''):
    try:
        int(ot_month)
    except ValueError:
        print('Wrong Month Format!')
    
    cur_year = int(date.today().strftime('%y'))
    cur_month = int(date.today().strftime('%m'))
    
    if not ot_month:
        ot_month = cur_month
    
    try:
        int(ot_month)
    except ValueError:
        print("'{0}' is Not Number!".format(ot_month))
    
    if int(ot_month) > 12 or int(ot_month) < 1:
        print('Wrong Month!')
        return None
    '''
    If current month is 2 ad ot_month input '6'
    Then program think this month is in the last year
    '''
    ot_month = int(ot_month)
    
    if 2 <= ot_month <= cur_month:
        last_month = ot_month - 1
        last_year = cur_year
    elif ot_month == 1:
        last_year = cur_year - 1
        last_month = ot_month + 11
    else:
        last_year = cur_year - 1
        last_month = ot_month - 1
        cur_year -= 1
    
    sta_date = '{:0>2d}{:0>2d}{:0>2d}'.format(last_year, last_month, 21)
    end_date = '{:0>2d}{:0>2d}{:0>2d}'.format(cur_year, ot_month, 20)
    ot_cycle = [str(sta_date), str(end_date)]
    
    return ot_cycle
    

def get_ot_in_cycle(start_stop_date=[]):
    sta_date = start_stop_date[0]
    stp_date = start_stop_date[1]
    ot_items_in_cycle = []
    
    for item in ot_items:
        if sta_date <= item[0] <= stp_date:
            if item not in ot_items_in_cycle:
                ot_items_in_cycle.append(item)
            
    return ot_items_in_cycle


def get_ot_date_list(start_stop_date=[]):
    if not start_stop_date:
        return None
    
    ot_lists = get_ot_in_cycle(start_stop_date)
    ot_set = set([])
    
    for ot_list in ot_lists:
        ot_set.add(ot_list[0])
    
    ot_date_list = list(ot_set)
    ot_date_list.sort()
    
    return ot_date_list


def get_one_day_ot_time(overtime_date):
    try:
        int(overtime_date)
    except ValueError:
        print('Wrong Time Report Date Format!')
    
    ot_month = overtime_date[2:4]
    ot_day = overtime_date[4:6]
    
    cur_month = int(date.today().strftime('%m'))
    ot_month = int(ot_month)
    if int(ot_day) > 20:
        ot_month_tmp = ot_month + 1
    else:
        ot_month_tmp = ot_month
    
    ot_cycle = get_ot_cycle(ot_month_tmp)
    ot_lists = get_ot_in_cycle(ot_cycle)
    ot_date_list = get_ot_date_list(ot_cycle)
    
    tmp_date = '20' + str('{:0>2d}'.format(ot_month)) + str(ot_day)
    ot_time_set = set([])
    for ot_list in ot_lists:
        if ot_list[0] == tmp_date:
            ot_time_set = ot_time_set | transform_rcms_date.sta_trans_to_set(ot_list[1], ot_list[2])
            # ot_time_set.add(transform_rcms_date.sta_trans_to_set(ot_list[1], ot_list[2]))
            '''
            time_point = 0
            while int(ot_list[1]) + time_point <= int(ot_list[2]):
                ot_time_set.add('{:0>3d}'.format(int(ot_list[1]) + time_point))
                time_point += 5
            '''
    
    ot_time_list = list(ot_time_set)
    ot_time_list.sort()
    
    print(ot_time_list)
    return ot_time_set




rcmsMenu.py
#!/usr/bin/env python

import time
from datetime import date
from os import system
from os import path
from functions import ibmPCommTLS
from functions import readConfig
from functions import ibmRetain
# from .functions import readConfig
# from .functions import ibmPCommTLS
# from . functions import ibmRetain

mainMenu = ['-------------IBM RCMS MENU-------------',
            '\033[0;90m'
            'Function',
            '\tCc. Current Config',
            '\tLl. Connect & Login',
            '\tPp. Print Current Page',
            '\tRr. Read Me',
            '\tMm. Main Menu',
            '\tHh. Help',
            '\tQq. Quit',
            '\033[0m'

            '\033[0;36m'
            'Call Read',
            '\tcr,[call no]     //eg. cr,p462yk9',
            '\tcr,[call no],w   //read the call every 60s waiting PARTS exist.',
            '\033[0m'

            '\033[0;35m'
            'Call Search&Read   //Only Support SDLNCN&SDLINC',
            '\tcs   //search for current user.',
            '\tcs,[cust no.],[notes id],[call type],[call queue]',
            '\033[0m'

            '\033[0;91m'
            'Call & Product Inventory',
            '\tci,[cust no.],[mach sn],[mach type]',
            '\tin,[mach type],[mach sn],[cust no.]',
            '\033[0m'

            '\033[0;33m'
            'SSR Information',
            '\tw,[ssr name],[ssr sn],[ssr retain id]',
            '\tNot Supported Yet!',
            '\033[0m'

            '\033[0;37m'
            'Time Report',
            '\tred //report list & summary for current user)',
            '\tred,b,s,w,t,e,d,[ssr SNs],[date from],[date to]',
            '\tb: backup to file;',
            '\ts: record list & summary (default date cutoff); ',
            '\tw: weekly summary (ignore date);',
            '\tt: UT summary for team(cut-off cycle);',
            '\te: team EPSB overlap verify;',
            '\td: one day detail;',
            '\tdate format: 200930; Only 1 date specified, deadline will be today.',
            '\tfor "t" or "e", if more than 1 ssr specified, config ignored.',
            '\033[0m'

            '\033[0;31m'
            'PMH',
            "\tz    //Search Current User's Opened PMH or last PMH been read before.",
            '\tz,args1,args2,args3,...',
            '\targs: [pmh no.: 49501 or 49501,000,672],',
            '\t[cust no.: 805543|0805543|00805543],',
            '\t[cust name: Gansu ICBC],',
            '\t[mach type: 1814],[mach sn: 78k15c8],',
            '\t[queue: SSRNW],[branch no.: "000"],',
            '\t[User Name: Zhang Zhi Jie],',
            '\t[pmh stat close|open]',
            '\tdr,[retain id] (read profile)',
            '\tce,[mach sn] (Create PMH)',
            '\tce,snap,[file dir]   //Analyze snap file & generate call. Coming...',
            '\tce,call no.  //rcms call no.; not supported yet.',
            '\033[0m\n']

subMenu4cr = ['Repeat',
              'export']

mainMenuIndex = ['C', 'L', 'P', 'R', 'Q', 'M', 'H']
cmdSet = ('CR', 'CS', 'CI', 'IN', 'RED', 'Z', 'DR', 'CE')
wait4Sec = 60
lastCallNo = ''
lastCallNoList = []
lastPMHNo = ''
autoLogin = True
host1 = ''
host2 = ''
ssrSN = ''
passwd = ''
cutFrom = ''
cutTo = ''
redSave2dir = path.abspath(path.join(path.dirname(__file__), 'log'))
teamSSRList = []
epsbSSRList2Check = []
rcmsConf = {}
retainConf = {}
cutoffDate = []

ibmPCommTLS.debugPrint('Verify config...', color='white')
try:
    rcmsConf = readConfig.getRCMSConf()
    cutoffDate = readConfig.getCutoffDate()
    teamSSRList = readConfig.getTeamSSRList()
    retainConf = readConfig.getRetainConf()
    for key in rcmsConf:
        print('\033[0;31m{:<12}: \033[0m{:.>25}'.format(key, rcmsConf[key]))

    print('\033[0;31mCutoff From : \033[0m\t{:.>23}'.format(cutoffDate[0]))
    print('\033[0;31mCutoff To   : \033[0m\t{:.>23}'.format(cutoffDate[1]))

    print('\033[0;31mTeam SSR List (first 6): \033[0m\n{}'.format(teamSSRList[0:6]))

    for key in retainConf:
        print('\033[0;31m{:<8}: \033[0m{:.>29}'.format(key, retainConf[key]))
except KeyError:
    print('\n')
    ibmPCommTLS.debugPrint('Key Error!, Check Config File.'
                           'Make sure keys include "host1", "host2", '
                           '"ssrSN", "teamSSRList", "epsbSSRList".')
    exit(127)

rcmsHost1 = rcmsConf['host1']
rcmsHost2 = rcmsConf['host2']
ssrSN = rcmsConf['ssrSN']
rcmsPwd = rcmsConf['passwd']
retainHost1 = retainConf['host1']
retainHost2 = retainConf['host2']
retainID = retainConf['retainID']
retainPwd = retainConf['passwd']

cutFrom = cutoffDate[0].replace('20', '', 1)
cutTo = cutoffDate[1].replace('20', '', 1)

for word in readConfig.getRCMSConf()['epsbSSRList'].split(','):
    epsbSSRList2Check.append(word.strip("'").strip('"').strip())

if not rcmsPwd:
    rcmsPwd = input('Enter your RCMS Password: ')
    if input('Confirm (Y/n): ').upper() != 'Y':
        exit(127)

if not retainPwd:
    retainPwd = input('Enter your Retain Password: ')
    if input('Confirm (Y/n): ').upper() != 'Y':
        exit(127)

ibmPCommTLS.debugPrint('Init IBM RCMS PComm Emulator...')
rcms = ibmPCommTLS.rcms(rcmsHost1, rcmsHost2, ssrSN, rcmsPwd)
ibmPCommTLS.debugPrint('RCMS Init Done.', color='green')
ibmPCommTLS.debugPrint('Init IBM Retain System...')
# *** need to modify
retain = ibmRetain.retain(retainHost1, retainHost2, retainID, retainPwd)
ibmPCommTLS.debugPrint('Retain Init Done.', color='green')
input('Enter Main Menu with Any Key.')
for menu in mainMenu:
    print(menu)

while True:
    # system('clear')
    prompt = input('CMD=> ').upper()
    while ' ,' in prompt or ', ' in prompt:
        prompt = prompt.replace(' ,', ',').replace(', ', ',')
    if not prompt.startswith(cmdSet) and prompt not in mainMenuIndex:
        ibmPCommTLS.debugPrint('Input not Permitted.', color='red')
        continue

    """
    # Auto login is embedded in object 'rcms'.
    if prompt not in mainMenuIndex:
        if rcms.currPageType() != rcms.pages['CMD=>']:
            ibmPCommTLS.debugPrint('SYSTEM ERROR! ENTER "L/l" TO LOGIN.', color='red')
            if autoLogin:
                ibmPCommTLS.debugPrint('Auto login to IBM RCMS...')
                rcms.login()
                rcms.printCurrScreen()
                ibmPCommTLS.debugPrint('Login done. Start jobs unfinished before.', color='green')
            else:
                input('\nPress Enter to be Continue!\n')
                continue
    """

    if prompt == 'C':
        print('\033[0;37mCurrent Config:\033[0m')
        for rc in rcmsConf:
            print('\033[0;31m{}: \033[0m{}'.format(rc.upper(), rcmsConf[rc]))
        for rc in retainConf:
            print('\033[0;31m{}: \033[0m{}'.format(rc.upper(), retainConf[rc]))
        print('\033[0;37mCurrent Cut-Off Day:\033[0m {}--{}'.format(cutFrom, cutTo))

    elif prompt == 'L':
        ibmPCommTLS.debugPrint('Connect to IBM RCMS host and login.')
        rcms.go2MainMenu()
        ibmPCommTLS.debugPrint('Connected, RCMS login done.')
        rcms.printCurrScreen()

    elif prompt == 'P':
        rcms.printCurrScreen()

    elif prompt == 'R':
        ibmPCommTLS.debugPrint('Nothing.', color='yellow')

    elif prompt == 'M' or prompt == 'H':
        system('clear')
        for menu in mainMenu:
            print(menu)

    elif prompt == 'Q':
        rcms.disconnect()
        rcms.quit()
        ibmPCommTLS.debugPrint('Quit', color='green')
        exit(0)

    # Call read by no.
    elif prompt.startswith('CR'):
        tmpList = prompt.split(',')
        if len(tmpList) == 2:
            callNo = tmpList[1]
            rcms.callRead(callNo)
            continue

        elif len(tmpList) == 3:
            callNo = tmpList[1]
            while True:
                callInfo = rcms.callRead(callNo)
                if '0000' in callInfo['action plan']:
                    ibmPCommTLS.debugPrint('Parts found, print call info.', color='green')
                    break
                if 'PERCALL' in callInfo['comments'] or 'CLS' in callInfo['st']:
                    ibmPCommTLS.debugPrint('Percall or Call closed, '
                                           'find parts in comment.', color='yellow')
                    break
                ibmPCommTLS.debugPrint('No part found, wait 30s for next loop.')
                time.sleep(wait4Sec)
            lastCallNo = callNo
            continue

        else:
            # ibmPCommTLS.debugPrint('Read Call No Error!', color='red')
            ibmPCommTLS.debugPrint('No Call No. or wrong args, read the last one.', color='red')
            rcms.callRead()

    # Call search.
    elif prompt.startswith('CS'):
        srchArgs = prompt.split(',')
        srchArgs.remove('CS')
        callNoList = rcms.callSrchPlus(srchArgs)
        for call in callNoList:
            print(call)
        prompt = input('(1-99) to Read, Enter back to menu: ')
        if prompt:
            try:
                rcms.callRead(callNoList[int(prompt) - 1][3:10])
            except ValueError:
                ibmPCommTLS.debugPrint('Value Error!')
            except IndexError:
                ibmPCommTLS.debugPrint('Index out of range!')

    # Call Inventory
    elif prompt.startswith('CI'):
        ciArgs = prompt.split(',')
        ciArgs.remove('CI')
        callInvList = rcms.callInquiryPlus(ciArgs)
        for ci in callInvList:
            print(ci)
        prompt = input('(1-99) to Read, Enter back to menu: ')
        if prompt:
            try:
                rcms.callRead(callInvList[int(prompt) - 1][3:10])
            except ValueError:
                ibmPCommTLS.debugPrint('Value Error!')
            except IndexError:
                ibmPCommTLS.debugPrint('Index out of range!')

    elif prompt.startswith('IN'):
        # inArgs = prompt.lstrip('IN,').split(',')
        inArgs = prompt.split(',')
        inArgs.remove('IN')
        proInvList = rcms.proInvPlus(inArgs)
        for pi in proInvList:
            print(pi)
        if len(proInvList) == 1:
            for pidtl in rcms.proInvDetail():
                print(pidtl)

    elif prompt.startswith('RED'):
        # redArgs = prompt.lstrip('RED,').split(',')
        redArgs = prompt.split(',')
        redArgs.remove('RED')
        rcms.redPlus(redArgs, cutFrom, cutTo, teamSSRList, epsbSSRList2Check)

    elif prompt.startswith('Z'):
        pmhArgs = prompt.split(',')
        pmhArgs.remove('Z')
        retain.readProPlus(pmhArgs)
        # print('No args, nothing got.')

    elif prompt.startswith('DR'):
        drArgs = prompt.split(',')
        drArgs.remove('DR')
        if not drArgs:
            drArgs.append('149715')
        print(drArgs)
        retain.displayProfile(drArgs[0])

    elif prompt.startswith('CE'):
        ceArgs = prompt.split(',')
        ceArgs.remove('CE')
        proInvList = rcms.proInvPlus(ceArgs)
        machDetail = ''
        if len(proInvList) == 0:
            input('Nothing found, please verify the machine serial number.')
            continue
        elif len(proInvList) != 1:
            for pi in proInvList:
                print(pi)
            prompt = input('More then 1 machine found, choose one: ')
            try:
                machDetail = proInvList[int(prompt)-1]
            except KeyError:
                print('Input error!')
        else:
            machDetail = proInvList[0]
        print(machDetail)
        retain.callEntryPlus(machDetail, retainConf['locQueue'])

